### PR TITLE
GetValueByNamespace extended to support retrieving values from maps a…

### PR DESCRIPTION
**Pull request by mkuculyma**
…nd arrays

Changes made in this PR:
- extended GetValueByNamespace
- unit tests updated - added cases of retrieving values from arrays and maps.

### CODE REVIEW

File: ns/namespace.go [hunk 1]
```diff
@@ -94,30 +94,85 @@ func FromCompositionTags(object interface{}, current string, namespace *[]string
...
+			if len(ns) == 1 {
+				return a[curr]
+			}
+			return GetValueByNamespace(a[curr], ns[1:])
+		}
+	case reflect.Ptr:
+		return GetStructValueByNamespace(object, ns)
+	case reflect.Struct:
+		return GetStructValueByNamespace(object, ns)
+	}
```
Comment (IzabellaRaulin) [inline ns/namespace.go]: I suggest adding `default` case to this switch with some log

File: ns/namespace.go [hunk 2]
```diff
@@ -94,30 +94,85 @@ func FromCompositionTags(object interface{}, current string, namespace *[]string
...
 // It iterates over fields recursively, checks tags until it finds leaf value.
-func GetValueByNamespace(object interface{}, ns []string) interface{} {
+func GetStructValueByNamespace(object interface{}, ns []string) interface{} {
 	// current level of namespace
 	current := ns[0]
+
 	fields, err := reflections.Fields(object)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Could not return fields for object{%v}\n", object)
+		fmt.Printf("Could not return fields for object{%v}\n", object)
```
Comment (IzabellaRaulin) [inline ns/namespace.go]: Why has it been changed to `Printf`?

Comment (mkuculyma) [inline ns/namespace.go]: fixed

File: ns/namespace.go (hunk 1 above)
Comment (mkuculyma) [inline ns/namespace.go]: fixed

File: ns/namespace.go [hunk 3]
```diff
@@ -94,12 +94,69 @@ func FromCompositionTags(object interface{}, current string, namespace *[]string
 		InspectEmptyContainers(AlwaysFalse))
 }
 
-// GetValueByNamespace returns value stored in struct composition.
+// GetValueByNamespace returns value stored in nested map, array or struct composition.
+func GetValueByNamespace(object interface{}, ns []string) interface{} {
+
+	if len(ns) == 0 {
+		fmt.Fprintf(os.Stderr, "Namespace length equal to zero\n")
```
Comment (lmroz) [inline ns/namespace.go]: Error handling should be done using second return argument, not just error console. Eventually you can just panic. It's better option than silently operating on incorrect data.

**Comment by lmroz** [comment]
I think this code cannot manage with cases when:
- On some level on struct you have a map or slice
- Value is a pointer to anything else than struct
Even if that gonna stay unimplemented at least you should leave `//TODO` or bug godoc comment.

File: ns/namespace.go (hunk 3 above)
Comment (marcin-krolik) [inline ns/namespace.go]: Changing signature of this function to introduce additional error handling would require changes in plugins which are using it. I would like it as is, plugin author can handle `nil` values which this function returns on failure.
As for `panic` this is really bad idea to crash whole plugin on problem with getting value for requested metric.

**Comment by mkuculyma** [comment]
Thanks, // TODO added

**Comment by IzabellaRaulin** [comment]
There is an issue about enhancement handle in `pointer` case https://github.com/intelsdi-x/snap-plugin-utilities/issues/34

**Pull request merged**